util/auto_tune_compressor.cc

```diff
@@ -0,0 +1,317 @@
+    return allcompressors_[choosen_compression_type][compression_level_ptr]
+        ->CompressBlock(uncompressed_data, compressed_output,
+                        out_compression_type, wa);
+  }
+  auto local_wa = static_cast<CostAwareWorkingArea*>(wa->get());
+  auto choosen_index = allcompressors_index_[Random::GetTLSInstance()->Uniform(
+      static_cast<int>(allcompressors_index_.size()))];
+  TEST_SYNC_POINT_CALLBACK(
+      "CostAwareCompressor::CompressBlock::SelectCompressionTypeAndLevel",
+      &choosen_index);
+  size_t choosen_compression_type = choosen_index.first;
+  size_t compresion_level_ptr = choosen_index.second;
+  return CompressBlockAndRecord(choosen_compression_type, compresion_level_ptr,
+                                uncompressed_data, compressed_output,
+                                out_compression_type, local_wa);
+}
+
+Compressor::ManagedWorkingArea CostAwareCompressor::ObtainWorkingArea() {
+  auto wrap_wa = allcompressors_.back().back()->ObtainWorkingArea();
+  auto wa = new CostAwareWorkingArea(std::move(wrap_wa));
+  // Create cost predictors for each compression type and level
+  wa->cost_predictors.reserve(compression_levels_.size());
+  for (size_t i = 0; i < compression_levels_.size(); i++) {
+    CompressionType type_ = static_cast<CompressionType>(i + 1);
+    if (compression_levels_[type_ - 1].size() == 0) {
+      wa->cost_predictors.emplace_back();
+      continue;
+    } else {
+      std::vector<IOCPUCostPredictor*> predictors_diff_levels;
+      predictors_diff_levels.reserve(compression_levels_[type_ - 1].size());
+      for (size_t j = 0; j < compression_levels_[type_ - 1].size(); j++) {
+        predictors_diff_levels.emplace_back(new IOCPUCostPredictor(10));
+      }
+      wa->cost_predictors.emplace_back(std::move(predictors_diff_levels));
+    }
+  }
+  return ManagedWorkingArea(wa, this);
+}
+void CostAwareCompressor::ReleaseWorkingArea(WorkingArea* wa) {
+  // remove all created cost predictors
+  for (auto& prdictors_diff_levels :
+       static_cast<CostAwareWorkingArea*>(wa)->cost_predictors) {
+    for (auto& predictor : prdictors_diff_levels) {
+      delete predictor;
+    }
+  }
+  delete static_cast<CostAwareWorkingArea*>(wa);
+}
+
+Status CostAwareCompressor::CompressBlockAndRecord(
+    size_t choosen_compression_type, size_t compression_level_ptr,
+    Slice uncompressed_data, std::string* compressed_output,
+    CompressionType* out_compression_type, CostAwareWorkingArea* wa) {
+  assert(choosen_compression_type < allcompressors_.size());
+  assert(compression_level_ptr <
+         allcompressors_[choosen_compression_type].size());
+  StopWatchCPUMicros timer(Env::Default()->GetSystemClock().get(), true);
+  Status status =
+      allcompressors_[choosen_compression_type][compression_level_ptr]
+          ->CompressBlock(uncompressed_data, compressed_output,
+                          out_compression_type, &(wa->wrapped));
```

Comment:
You probably want to scope the timer with CompressBlock call using {} so the timer only measure CompressBlock() call